File: pronto2broadlink.js
// Javascript port of https://gist.githubusercontent.com/appden/42d5272bf128125b019c45bc2ed3311f/raw/bdede927b231933df0c1d6d47dcd140d466d9484/pronto2broadlink.py
// Discovered at https://www.reddit.com/r/homeautomation/comments/7m7ddv/broadlink_ir_database/dru77am/
// More protocol documentation at https://github.com/mjg59/python-broadlink/blob/master/protocol.md
// Run standalone with:
// node pronto2broadlink.js "<pronto code>"

function pronto2lirc(pronto) {
  for (var codes = [], i = 0; i < pronto.length - 1; i += 2) {
    codes.push((pronto[i] << 8) + pronto[i + 1])
  }

  if (codes[0] !== 0) {
    raise('Pronto code should start with 0000')
  }
  if (codes.length !== 4 + 2 * (codes[2] + codes[3])) {
    raise('Number of pulse widths does not match the preamble')
  }

  period = 1 / (codes[1] * 0.241246)
  return codes.slice(4).map(c => {
    return Math.round(c / period)
  })
}

function lirc2broadlink(pulses, repeats) {
  var payload = broadlinkEncodePayload(pulses)

  var lengthBytes = bigEndianToLittleEndian(payload, 2);

  var packet = []
  packet.push(0x26, repeats)
  packet.push(...lengthBytes)
  packet.push(...payload, 0x0d, 0x05)

  remainder = (packet.length + 4) % 16  // rm.send_data() adds 4-byte header (02 00 00 00)
  if (remainder > 0) {
    packet.push(...Array(16 - remainder).fill(0))
  }

  return packet

}

function bigEndianToLittleEndian(payload, nbytes) {
  var lengthHexStrBE = pad(payload.length.toString(16), nbytes * 2); // Big endian (default)
  var lengthHexStrLE = lengthHexStrBE.split().reverse().join(""); // Little endian

  for (var leBytes = [], i = 0; i < lengthHexStrLE.length; i += 2) {
    leBytes.push(parseInt(lengthHexStrLE.substr(i, 2),16))
  }

  return leBytes.reverse();
}

function broadlinkEncodePayload(pulses) {
  for (var payload = [], i = 0; i < pulses.length; i++) {
    var pulse = Math.floor(pulses[i] * 269 / 8192)

    if (pulse < 256) {
      payload.push(pulse)
    } else {
      payload.push(0x00, (pulse & 0xFF00) >> 8, pulse & 0x00FF)
    }
  }

  return payload
}

function b64encode(packet) {
  return Buffer.from(packet).toString('base64')
}

function pad(num, size) {
  var s = num + "";
  while (s.length < size) s = "0" + s;
  return s;
}

function hexToBytes(hexRaw) {
  hex = hexRaw.split(' ').join('')
  for (var bytes = [], c = 0; c < hex.length; c += 2)
    bytes.push(parseInt(hex.substr(c, 2), 16));
  return bytes;
}

function printArray(arr) {
  console.log(`[${arr.join(', ')}]`)
}

if (require.main === module) {
  process.argv.slice(2).forEach(code => {
    pronto = hexToBytes(code)
    console.log('ProntoBytes:')
    printArray(pronto.map(b => `'${b.toString(16)}'`))

    pulses = pronto2lirc(pronto)
    console.log('Pulses:')
    printArray(pulses)

    packet = lirc2broadlink(pulses, 0x00)
    console.log('Packet:')
    printArray(packet.map(b => `'${b.toString(16)}'`))

    var b64Packet = b64encode(packet)
    console.log('Base64 packet:')
    console.log(b64Packet)
  })
}

File: pronto2broadlink.py
# Python 3 port of https://gist.githubusercontent.com/appden/42d5272bf128125b019c45bc2ed3311f/raw/bdede927b231933df0c1d6d47dcd140d466d9484/pronto2broadlink.py
# Discovered at https://www.reddit.com/r/homeautomation/comments/7m7ddv/broadlink_ir_database/dru77am/
# More protocol documentation at https://github.com/mjg59/python-broadlink/blob/master/protocol.md
# Run standalone with:
# python pronto2broadlink.py "<pronto code>"

import struct
import binascii
import hexdump as hexd
from base64 import b64encode

def pronto2lirc(pronto):
    codes = [int(binascii.hexlify(pronto[i:i+2]), 16) for i in range(0, len(pronto), 2)]

    if codes[0]:
        raise ValueError('Pronto code should start with 0000')
    if len(codes) != 4 + 2 * (codes[2] + codes[3]):
        raise ValueError('Number of pulse widths does not match the preamble')

    frequency = 1 / (codes[1] * 0.241246)
    return [int(round(code / frequency)) for code in codes[4:]]

def lirc2broadlink(pulses):
    array = bytearray()

    for pulse in pulses:
        pulse = (int)(pulse * 269 / 8192)  # Pulse lengths in 2^-15s units

        if pulse < 256:
            array += bytearray(struct.pack('>B', pulse))  # big endian (1-byte)
        else:
            array += bytearray([0x00])  # indicate next number is 2-bytes
            array += bytearray(struct.pack('>H', pulse))  # big endian (2-bytes)
    print(len(array))
    packet = bytearray([0x26, 0x01])  # 0x26 = IR, 0x00 = no repeats
    packet += bytearray(struct.pack('<H', len(array)))  # little endian byte count
    packet += array
    packet += bytearray([0x0d, 0x05])  # IR terminator

    # Add 0s to make ultimate packet size a multiple of 16 for 128-bit AES encryption.
    remainder = (len(packet) + 4) % 16  # rm.send_data() adds 4-byte header (02 00 00 00)
    print(len(packet))
    if remainder:
        packet += bytearray(16 - remainder)

    return packet

if __name__ == '__main__':
    import sys

    for code in sys.argv[1:]:
        pronto = bytearray.fromhex(code)
        print('ProntoBytes:')
        print(pronto)
        print([hex(p)[2:] for p in pronto])

        pulses = pronto2lirc(pronto)
        print('Pulses:')
        print(pulses)

        packet = lirc2broadlink(pulses)
        print('Packet:')
        print(packet)
        print([hex(p)[2:] for p in packet])

        print('Hex packet:')
        hexPacket = hexd.dump(packet)

        b64Packet = b64encode(packet)
        print('Base64 packet:')
        print(b64Packet.decode('utf-8'))